interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp

```diff
@@ -0,0 +1,968 @@
+
+  std::string GetFullName() const {
+    llvm::SmallString<512> Vec(m_Path);
+    llvm::sys::path::append(Vec, llvm::StringRef(m_LibName));
+    return Vec.str().str();
+  }
+
+  void AddBloom(llvm::StringRef symbol) {
+    m_Filter.AddHash(GNUHash(symbol));
+  }
+
+  llvm::StringRef AddSymbol(const std::string& symbol) {
+    auto it = m_Symbols.insert(symbol);
+    return it.first->getKey();
+  }
+
+  bool isBloomFilterEmpty() const {
+    assert(m_Filter.m_IsInitialized && "Bloom filter not initialized!");
+    return m_Filter.m_SymbolsCount == 0;
+  }
+
+  void InitializeBloomFilter(uint32_t newSymbolsCount) {
+    assert(isBloomFilterEmpty() && "Cannot re-initialize non-empty filter!");
+    assert(!m_Filter.m_IsInitialized &&
+           "Cannot re-initialize non-empty filter!");
+    m_Filter.m_IsInitialized = true;
+    m_Filter.ResizeTable(newSymbolsCount);
+  }
+
+  bool MayExistSymbol(uint32_t hash) const {
+    // The library had no symbols and the bloom filter is empty.
+    if (isBloomFilterEmpty())
+      return false;
+
+    return m_Filter.TestHash(hash);
+  }
+
+  bool ExistSymbol(llvm::StringRef symbol) const {
+    return m_Symbols.find(symbol) != m_Symbols.end();
+  }
+};
+
+
+/// A helper class keeping track of loaded libraries. It implements a fast
+/// search O(1) while keeping deterministic iterability in a memory efficient
+/// way. The underlying set uses a custom for better efficiency given the
+/// specific problem where the library names (m_LibName) are relatively short
+/// strings and the base paths (m_Path) are repetitive long strings.
+class LibraryPaths {
+  struct LibraryPathHashFn {
+    size_t operator()(const LibraryPath& item) const {
+      return std::hash<size_t>()(item.m_Path.length()) ^
+        std::hash<std::string>()(item.m_LibName);
+    }
+  };
+
+  std::vector<const LibraryPath*> m_Libs;
+  std::unordered_set<LibraryPath, LibraryPathHashFn> m_LibsH;
+public:
+  const LibraryPath& RegisterLib(const BasePath& BasePath,
+                                 const std::string& LibFileName) {
+    auto it = m_LibsH.insert(LibraryPath(BasePath, LibFileName));
+    const LibraryPath& result = *it.first;
+    if (it.second)
+      m_Libs.push_back(&result);
+    return result;
+  }
+
+  void UnregisterLib(const LibraryPath& Lib) {
+    auto found = m_LibsH.find(Lib);
+    if (found == m_LibsH.end())
+      return;
+
+    m_Libs.erase(std::find(m_Libs.begin(), m_Libs.end(), &*found));
+    m_LibsH.erase(found);
+  }
+
+  size_t size() {
+    assert(m_Libs.size() == m_LibsH.size());
+    return m_Libs.size();
+  }
+
+  const std::vector<const LibraryPath*>& GetLibraries() {
+    return m_Libs;
+  }
+};
+
+
+Dyld::Dyld(const cling::DynamicLibraryManager& DLM)
+  : m_DynamicLibraryManager(DLM) {
+  m_Libraries = llvm::make_unique<LibraryPaths>();
+  m_SysLibraries = llvm::make_unique<LibraryPaths>();
+}
+
+Dyld::~Dyld() {}
+
+static bool s_IsDyldInitialized = false;
+static llvm::StringSet<> s_LibBlacklist;
```

Comment:
That would mean a dyld per dynamic library manager. That should be doable, not sure if we want this in this PR as it will take us a while to make that sort of refactoring.

Let's aim for merging and then fixing this after the release?